### PR TITLE
feat(resources): support static `$au` property for definition

### LINKED ISSUE #1749: [refactor] simplify resources system
## 🔦 Context

At the moment, resources are registered into containers via their corresponding resource definitions. This means by the time of registration, resource definitions should be ready, which puts some pressure of the preparation of its data/configuration. We have had cases where one decorator needs to be aware of another, which makes them order sensitive, or poke into internal of others. Also, the reliance on metadata requires application to have to import from Aurelia, which prohibits the Aurelialess way of authoring component that exists in v1.

Consider a way to simply the resource system, so that it doesn't require the definition to be constructed too early. This will also allow renderer to resolve and cache resource definition independently per application, rather than statically on the classes. And this should also help simplify generated code for our tooling, going forward with precompilation.

## 💻 Code Sample

```ts
// example of declaring a custom element with static `$res`
export class MyElement {
  static $res = 'element'
}
```

### PR DESCRIPTION
**Pull request by bigopon**
## 📖 Description

Add the ability to declare a resource via a static `$au` property, like the following example:
```ts
export class MyElement {
  static $au = {
    type: 'custom-element',
    name: 'my-el'
  }
  ...
}
```
This helps with avoiding having to use decorator if one prefers to.

Additionally, tweak the typings of bindables in the custom element/attribute definition, so the `name` property is no longer required in object form, and object form is allowed in array usage, like the following examples:
```ts
bindables: [
  'prop',
  { name: 'prop' , ... }
]
// or
bindables: {
  prop: true,
  prop: { ... }
}
```

### 🎫 Issues

Closes #1749 

cc @Sayan751 @fkleuver

### CODE REVIEW

**Comment by bigopon** [comment]
@Sayan751 @3cp we can adjust the code generated by the convention plugin to use static property `$au` instead, so that we don't have to deal with decorator compatibility of build tools. The static convention works with independent decorators so applications don't need to make any changes, for example:
```ts
class MyEl {
  static $au = { ... }
  @bindable value
}
```
The above works fine.

**Comment by 3cp** [comment]
This will simplify conventions greatly, as it removes all the import of decorators!

**Review by Sayan751 — COMMENTED** [review]
Hey @bigopon! Great work! Just got one question.

File: packages/runtime-html/src/resources/custom-element.ts [hunk 1]
```diff
@@ -365,7 +373,7 @@ export class CustomElementDefinition<C extends Constructable = Constructable> im
     // todo: warn if alreay has key
     if (!container.has(key, false)) {
       container.register(
-        container.has($Type, false) ? null : transientRegistration($Type, $Type),
+        container.has($Type, false) ? null : singletonRegistration($Type, $Type),
```
[the remark this answers is not in the document]
Comment (bigopon) [reply packages/runtime-html/src/resources/custom-element.ts:376]: It is. This is just for registration, I don't think we even need this, as resources registration should be via key only. We are not getting resources through `container.get` (traversal behavior), so this shouldn't matter.

File: packages/runtime-html/src/compiler/template-compiler.ts [hunk 2]
```diff
@@ -1725,11 +1722,11 @@ class CompilationContext {
     return new CompilationContext(this.def, this.c, this.ci, this, this.root, instructions);
   }
 
-  // todo: ideally binding command shouldn't have to be cached
-  // it can just be a singleton where it' retrieved
-  // the resources semantic should be defined by the resource itself,
-  // rather than baked in the container
-  private readonly _commands: Record<string, BindingCommandInstance | null | undefined> = createLookup();
+  // // todo: ideally binding command shouldn't have to be cached
```
[the remark this answers is not in the document]
Comment (bigopon) [reply packages/runtime-html/src/compiler/template-compiler.ts:1725]: Ugh, I thought I removed this change. Planned to do it later when extracting the compiler. Will clean this up thanks @Sayan751

[the remark this answers is not in the document]
Comment (Sayan751) [reply packages/runtime-html/src/compiler/template-compiler.ts:1725]: No worries. I also can clean it up in my PR.

[the remark this answers is not in the document]
Comment (bigopon) [reply packages/runtime-html/src/compiler/template-compiler.ts:1725]: Thatd be good too, thanks!

**Pull request merged**